lib/Transforms/HandshakeOptimization.cpp

```diff
@@ -0,0 +1,629 @@
+#include "dynamatic/Transforms/HandshakeOptimization.h"
+#include "dynamatic/Dialect/Handshake/HandshakeCanonicalize.h"
+#include "dynamatic/Dialect/Handshake/HandshakeOps.h"
+#include "dynamatic/Support/LLVM.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/Value.h"
+#include "mlir/IR/ValueRange.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+#include <vector>
+using namespace mlir;
+using namespace dynamatic;
+
+namespace {
+
+struct RemoveBranchMUXPairs
+    : public OpRewritePattern<handshake::ConditionalBranchOp> {
+  using OpRewritePattern<handshake::ConditionalBranchOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(handshake::ConditionalBranchOp condBranchOp,
+                                PatternRewriter &rewriter) const override {
+    mlir::Operation *useOwner1 = nullptr;
+    mlir::Operation *useOwner2 = nullptr;
+
+    auto it = condBranchOp.getTrueResult().getUsers().begin();
+    if (it != condBranchOp.getTrueResult().getUsers().end()) {
+      useOwner1 = *it;
+    }
+
+    it = condBranchOp.getFalseResult().getUsers().begin();
+    if (it != condBranchOp.getFalseResult().getUsers().end()) {
+      useOwner2 = *it;
+    }
+
+    if (isa_and_nonnull<handshake::MuxOp>(useOwner1) and
+        useOwner1 == useOwner2) {
+      handshake::MuxOp muxOp = mlir::cast<handshake::MuxOp>(useOwner1);
+      mlir::Value conditionBr = condBranchOp.getConditionOperand();
+      mlir::Value selectMux = muxOp.getSelectOperand();
+
+      if (conditionBr.getDefiningOp() == selectMux.getDefiningOp() and isa_and_nonnull<handshake::ForkOp>(conditionBr.getDefiningOp())) {
+        mlir::Value dataOperand = condBranchOp.getDataOperand();
+        rewriter.create<handshake::SinkOp>(muxOp->getLoc(), selectMux);
```

Comment:
You need to set the insertion point somewhere before you create operations, otherwise they may not be inserted where you want.